File: Algorithm/leetcode_560.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# File              : leetcode_560.py
# Author            : Andy
# Date              : 30.07.2022
# Last Modified Date: 30.07.2022
# Last Modified By  : Andy


class Solution:
    def subarraySum(self, nums, k) :
        n = lens(nums)
        ans = 0 # 最终要返回的结果计数，初始化为0
        dic = {0:1} # 前面结果为0的子数组次数为1
        sum_i = 0 # 前缀和的值初始化为0 

        for i in range(n):
            # 前缀和，累加
            sum_i += nums[i]
            # 如果要找的部分前缀的值已经出现过，那么通过当前的前缀和与这部分前缀相减就可以得到这个k值，
            # 只需要将这部分前缀出现的次数加上即可,
            # 如果没有出现，继续统计，把当前的前缀和的值存入字典
            # 如果当前前缀和的值还没有出现，那么置为1，即首次出现
            if sum_i - k in dic:
                ans += dic.get(sum_i-k)
            dic[sum_i] = dic.get(sum_i,0) + 1
        return ans

File: Algorithm/leetcode_560.go
/**
 * File              : leetcode_560.go
 * Author            : Andy
 * Date              : 30.07.2022
 * Last Modified Date: 30.07.2022
 * Last Modified By  : Andy
 */

// 主体思路：记录前缀和，以及前缀和中每个数字出现的次数
// 如果当前前缀和-k的值在map中，那么说明可能通过两个前缀和相减得到指定的k
// 而记录了这各前缀和的次数，变避免了遍历

func subarraySum(nums []int, k int) int{
    ans := 0
    val dict = make(map[int]int{0:1})
    sums := 0
    for let i:=0;i < len(nums);i++{
        // 前缀和相加
        sums += nums[i]
        _, ok := dict[sums-k]
        if ok {
            ans += val
        }
        s_val, s_ok := dict[sums]
            if s_ok{
                dict[sums] = s_val + 1
            }else{
                dict[sums] = 1
            }
    }
    return ans
}
